platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/render/RenderTestActivity.java

```diff
@@ -0,0 +1,156 @@
+      renderTest(renderTestDefinition, renderTestDefinitions.length);
+    }
+  }
+
+  private void renderTest(final RenderTestDefinition renderTestDefinition, final int testSize) {
+    MapSnapshotter mapSnapshotter = new MapSnapshotter(this, renderTestDefinition.toOptions());
+    mapSnapshotterList.add(mapSnapshotter);
+    mapSnapshotter.start(new MapboxMap.SnapshotReadyCallback() {
+      @Override
+      public void onSnapshotReady(Bitmap snapshot) {
+        imageView.setImageBitmap(snapshot);
+        renderResultMap.put(renderTestDefinition, snapshot);
+        if (renderResultMap.size() == testSize) {
+          writeResultsToDisk();
+          finishedRendering = true;
+        }
+      }
+    });
+  }
+
+  private void writeResultsToDisk() {
+    if (isExternalStorageWritable()) {
+      try {
+        File testResultDir = createTestResultRootFolder();
+        String basePath = testResultDir.getAbsolutePath();
+
+        for (Map.Entry<RenderTestDefinition, Bitmap> testResult : renderResultMap.entrySet()) {
+          String testName = testResult.getKey().getName();
+          String testDir = createTestDirectory(basePath, testName);
+          writeTestResultToDisk(testDir, testResult.getValue());
+        }
+      } catch (Exception exception) {
+        Timber.e(exception);
+      }
+    }
+  }
+
+  private File createTestResultRootFolder() {
+    File testResultDir = new File(Environment.getExternalStorageDirectory() + "/mapbox");
+    if (testResultDir.exists()) {
+      // cleanup old files
+      deleteRecursive(testResultDir);
+    }
+
+    if (!testResultDir.mkdirs()) {
+      throw new RuntimeException("can't create root test directory");
+    }
+    return testResultDir;
+  }
+
+  private void deleteRecursive(File fileOrDirectory) {
+    if (fileOrDirectory.isDirectory()) {
+      for (File child : fileOrDirectory.listFiles()) {
+        deleteRecursive(child);
+      }
+    }
+
+    if (!fileOrDirectory.delete()) {
+      throw new RuntimeException("can't delete directory");
+    }
+  }
+
+  private String createTestDirectory(String basePath, String testName) {
+    File testDir = new File(basePath + "/" + testName);
+    if (!testDir.exists()) {
+      if (!testDir.mkdir()) {
+        throw new RuntimeException("can't create sub directory for " + testName);
+      }
+    }
+    return testDir.getAbsolutePath();
+  }
+
+  private void writeTestResultToDisk(String testPath, Bitmap testResult) {
+    try {
+      String filePath = testPath + "/actual.png";
+      FileOutputStream out = new FileOutputStream(filePath);
+      testResult.compress(Bitmap.CompressFormat.PNG, 100, out);
+      out.flush();
+      out.close();
+    } catch (Exception exception) {
+      throw new RuntimeException(exception);
+    }
+  }
+
+  @Override
+  protected void onStop() {
+    super.onStop();
+    for (MapSnapshotter snapshotter : mapSnapshotterList) {
+      snapshotter.cancel();
```

Comment:
Should not be strictly needed anymore since the snapshotters are deletable by the gc collector thread now. Anything specific you run into?

Reply:
will do some additional tests, I remember hitting a freeze but could be fixed in meantime.